arxiv/base/urls/links.py

```diff
@@ -302,11 +302,44 @@ def _compose_list_of_funcs(fv: List[Callable_Linker]) -> Callable_Linker:
     return reduce(lambda f, g: lambda x: f(g(x)), fv[1:], fv[0])
 
 
+# We don't want to call local() at package import time because that would be
+# only on one of the WSGI theads.
+_t_local_linkers = None
+
+
+def _deferred_thread_local_linker_of_kind(kind: str) -> Callable_Linker:
+    # Bleach linkifiers are not thread safe. This thread localizes them.
+    #
+    # This will return a Callable that creates the bleach linkifier at
+    # only when called with the string to linkify. This is so the bleach
+    # linkifier can be put into a thead.local of each individual WSGI thread.
+
+    def deferred(instr: str) -> str:
+        # This must not be called while the app is starting up, only when the
+        # linkiers are being used in threads. So the function is created/cached
+        # at call time.
+        global _t_local_linkers
```

Comment:
L321-328 could be something like:

```python
if linkers not in g:
    g.linkers = {}
if kind not in g.linkers:
    g.linkers[kind] = _get_linker_of_kind(kind)

return g.linkers[kind]
```

Reply:
It seems like urlize is just a utility to convert strings to other strings. It is kind of distressing that it then would be dependent on Flask and Werkzeurg. Maybe the setup to do the late creation of the bleach objects should be up in the Flask factory instead of in links.py?

Reply:
> It is kind of distressing that it then would be dependent on Flask and Werkzeurg.

The ``arxiv.base.urls`` module is already [dependent on Werkzeug and Flask](https://github.com/arXiv/arxiv-base/blob/master/arxiv/base/urls/__init__.py#L64), so this doesn't introduce a new dependency.

> Maybe the setup to do the late creation of the bleach objects should be up in the Flask factory instead of in links.py?

Not sure that I follow. Do you mean initializing ``g.linkers`` in [``arxiv.base.Base``](https://github.com/arXiv/arxiv-base/blob/master/arxiv/base/__init__.py#L38)?

The point of Flask's proxy objects is to provide thread- and context-safe extensibility for applications without having to wire everything up tightly together. This use-case is precisely what ``g`` is designed for.